src/provider/model/Model.ts

```diff
@@ -166,33 +165,63 @@ export class Model {
             // Return a connection once we have one, or timeout when the application is mature
             return Promise.race([
                 deferredPromise.promise.then((connection) => [connection]),
-                this.getOrCreateLiveApp(appInfo).then((liveApp) => liveApp.waitForAppMature().then(() => [], () => []))
+                liveApp.waitForAppMature().then(() => [], () => [])
             ]).then((connection) => {
                 slot.remove();
                 return connection;
             });
         }
     }
 
-    public async getOrCreateLiveApp(appInfo: Application): Promise<LiveApp> {
-        const uuid = AppDirectory.getUuidFromApp(appInfo);
+    public async getOrCreateLiveAppByName(name: AppName): Promise<LiveApp> {
+        return this.getOrCreateLiveApp<never>(
+            name,
+            // Either find an app with this name
+            (candidateAppInfo) => candidateAppInfo.name === name,
+            // Or get it from the app directory
+            () => this._directory.getAppByName(name),
+            // We have no additional criteria for this application
+            () => undefined,
+            () => undefined
+        );
+    }
 
-        if (this._liveAppsByUuid[uuid]) {
-            return this._liveAppsByUuid[uuid];
-        } else {
-            const deferredPromise = new DeferredPromise<LiveApp>();
+    public async getOrCreateLiveAppByNameForIntent(name: AppName, intentType: string, contextType: string): Promise<LiveApp | 'does-not-support-intent'> {
+        return this.getOrCreateLiveApp(
+            name,
+            // Either find an app with this name
+            (candidateAppInfo) => candidateAppInfo.name === name,
+            // Or get it from the app directory
+            () => this._directory.getAppByName(name),
+            // Additionally require that this application supports the given intent and context
+            (liveApp) => {
+                const hasIntentListener = liveApp.connections.some((connection) => connection.hasIntentListener(intentType));
+
+                if (hasIntentListener && AppDirectory.mightAppSupportIntent(liveApp.appInfo!, intentType, contextType)) {
+                    return undefined;
+                }
 
-            const slot = this._environment.onApplicationCreated.add((identity: Identity, liveApp: LiveApp) => {
-                if (identity.uuid === uuid) {
-                    slot.remove();
-                    deferredPromise.resolve(liveApp);
+                if (!liveApp.mature && AppDirectory.shouldAppSupportIntent(liveApp.appInfo!, intentType, contextType)) {
+                    return undefined;
                 }
-            });
 
-            this._environment.createApplication(appInfo);
+                return 'does-not-support-intent';
```

Comment:
Does this want to be a const?

Reply:
My feeling is no. the type system enforces that I can't type some arbitrary string here, and with stuff like `Promise<LiveApp | 'does-not-support-intent'>`, you have to use a string literal anyway, you can't extract that into a const